src/parser.rs

```diff
@@ -27,164 +29,149 @@ impl<'a, T: Read> Parser<T> {
     /// values you can call this multiple times.  If the reader is not yet
     /// ready this will block.
     pub fn parse_value(&mut self) -> RedisResult<Value> {
-        let b = try!(self.read_byte());
-        match b as char {
-            '+' => self.parse_status(),
-            ':' => self.parse_int(),
-            '$' => self.parse_data(),
-            '*' => self.parse_bulk(),
-            '-' => self.parse_error(),
-            _ => fail!((ErrorKind::ResponseError, "Invalid response when parsing value")),
+        let mut buf = vec![];
+        match self.buf_reader.read_to_end(&mut buf) {
```

Comment:
reading a stream to end will wait for an EOF, meaning for the connection to close.
That's definitely not what you want here.
You would want to read as much as possible, then try to parse it and if necessary read more of it.